### ISSUE TITLE
uiMainStep does not work with gtk backend

### ISSUE THREAD

**Issue by gnarz** [andlabs/libui#95]
it always returns false. Some investigation shows that this is because the function gtk_main_iteration_do(), which is used to implement this, needs a running main loop, and that is only created when calling gtk_main() (see https://git.gnome.org/browse/gtk+/tree/gtk/gtkmain.c?h=3.18.9). So i see no easy way to fix this... maybe an idle callback would be a prudent replacement, that should be available on all platforms...

**Comment by andlabs** [comment]
I could probably use one of the other main loop functions here. Would need to see the original PR for this feature to see how they did it.

**Comment by gnarz** [comment]
I've checked the original PR, they do it like you do. As the main loops are only managed from within gtk_main(), there seems to be no way to implement uiMainStep without major hackery... the gtk_main_iteration*() functions seem to be intended to be used from idle callbacks to update the gui in case of long running computations.

**Comment by andlabs** [comment]
I wonder if a workaround could be to call uiQueueMain() with your real main loop and call uiMain() anyway. Whether the uiQueueMain() handler needs to call uiQuit() at the end depends on if I make uiMainStep() honor the outermost main loop or not.

**Comment by gnarz** [comment]
hmmm... from what I see, we're entering multithreaded territory with uiQueueMain, at least on windows and mac. Lua can't do that, so no, this will not work. For my use case, at least. As mentioned before, an idle callback might be a solution for the single threaded case.

**Comment by andlabs** [comment]
No mutlithreading here. QueueMain basically does

```
if (there are no messages waiting)
    run our queued callback();
else
    process events();
```

**Comment by andlabs** [comment]
Okay now that I'm more awake...

No, `uiQueueMain()` does not involve multiple threads. In fact, that's part of the name: it *queues* something to run on the *main* thread. It's provided for multithreaded applications, yes, but it doesn't have to be used by them. It can certainly be used by the main thread as well.

What QueueMain waits for is for no events to be pending. This is done with `gdk_threads_add_idle()` on GTK+ and `dispatch_async()` on OS X. On Windows, this is done with `PostMessage(); the posted message will come after any other (hihg-priority) message currently in the message queue.

So what this `uiMainSteps()` would do is just queue a function to run on the main loop. This function would never return, and would run its own main loop instead. I would have to add extra code to tell the real main loop to quit when the fake one has to quit as well, but that's no big deal.

And in fact, this is only necessary on GTK+; the other platforms can just call the main steps function directly.

Hopefully this explanation is sufficient.

**Comment by gnarz** [comment]
yes it is, thanks. I was / am not so sure about dispatch_async, because the man page seems to imply that it may be using threaded execution. I am not that firm in Mac OS programming, so I am probably wrong here (or this behaviour is "fixed" by using dispatch_get_main_queue() ). So, when I return from the function queued with uiQueueMain, will it automatically be called again when there are no events pening, or is it a one shot affair and I would have to re-register the function to have it called again?

... it is totally weird to be discussing this in the issues section, isn't there a forum or something?

**Comment by andlabs** [comment]
`uiQueueMain()` is one-shot, but you don't have to return. What I was suggesting was something like

```
void queued(void *data)
{
    while (uiMainStep(0)) {
        /* do your stuff here */
    }
}
```

and simply don't let the main loop run at all.

`dispatch_get_main_queue()` is the magic, yes. OS X doesn't allow you to do GUI stuff across threads, so if `dispatch_async()` didn't honor the queue thread, it wouldn't be useful. See also `dispatch_after()`.

**Comment by gnarz** [comment]
thanks a lot. I will try this after my vacation, but this also is the idle callback, if used properly. From what I've seen, your approach should also work with mac and windows, although not necessary there.

**Comment by andlabs** [comment]
Okay it turns out `uiMainStep()` never worked anyway, as on OS X `[NSApp isRunning]` returns `NO` until we call `[NSApp run]`. `uiMainSteps()` might be necessary, unless there's a workaround involving NSRunLoop...

**Comment by zevv** [comment]
So, this is getting messy. It's hard to make applications that do real work though, without some way of integrating the main loop with handling other events . (Unless you want to resort to threads, which I personally am no fan of because I long ago found out I'm not smart enough to make good code when threads are involved)

Idle functions or periodic polls are not nice either, because nobody wants to make applications which are CPU hogs.

Though problem here.

**Comment by andlabs** [comment]
Exactly. I would like to be able to pass in an array of file descriptors or something to poll with, because I'm not sure if the other way around is feasible either...

Unrelated: How exactly would you write code that uses this uiMainStepSelect()? Would it look something like this?

```c
for (;;) {
    int what = uiMainStepSelect(FDs, timeout);
    if (what == 0) break; // uiQuit()
    if (what == 1) continue; // window system message
    what -= 2;
    /* process FDs[what] */
}
```

**Comment by zevv** [comment]
* On 2016-06-13 16:01:32 +0200, Pietro Gagliardi wrote:
 
> Exactly. I would like to be able to pass in an array of file descriptors or something to poll with, because I'm not sure if the other way around is feasible either...
> 
> Unrelated: How exactly would you write code that uses this uiMainStepSelect()? Would it look something like this?
> 
> ```c
> for (;;) {
>     int what = uiMainStepSelect(FDs, timeout);
>     if (what == 0) break; // uiQuit()
>     if (what == 1) continue; // window system message
>     what -= 2;
>     /* process FDs[what] */
> }
> ```

Yes, something like that would do. If you want to be unix-ish, I guess
you best mimic the poll() system call.


```
fd_set fds[2] = {
	{ .fd = 0, .events = POLLIN },
	{ .fd = 13, .events = POLLERR },
};

for(;;) {
	int r = uiMainStepSelect(&fds, 2, timeout);

	if(r == 0) {
		/* timeout occured */
	} else if(r > 0) {
		/* one or more fd's have events */
	} else {
		/* returning -1 could mean quit ? */
	}
}
```

The semantics of returning -1 might not be very nice, maybe there could
be a function that the user polls in the main loop to indicate that the
app should quit?


```
while(!uiShouldQuit()) {
	...
}
```

-- 
:wq
^X^Cy^K^X^C^C^C^C

**Comment by daurnimator** [comment]
> Some investigation shows that this is because the function gtk_main_iteration_do(), which is used to implement this, needs a running main loop, and that is only created when calling gtk_main() (see https://git.gnome.org/browse/gtk+/tree/gtk/gtkmain.c?h=3.18.9). So i see no easy way to fix this... maybe an idle callback would be a prudent replacement, that should be available on all platforms...

There is no reason we can't call `g_main_loop_new` ourselves.

Then `uiMain` is a call to `g_main_loop_run ` and `uiMainStep` is a call to `g_main_context_iteration`.

**Comment by andlabs** [comment]
Actually looking at the GTK+ source code again, I wonder what would happen if I did this

```diff
diff --git a/unix/main.c b/unix/main.c
index bfd0544..0269d86 100644
--- a/unix/main.c
+++ b/unix/main.c
@@ -41,7 +41,8 @@ int uiMainStep(int wait)
 	block = FALSE;
 	if (wait)
 		block = TRUE;
-	return gtk_main_iteration_do(block) == FALSE;
+	gtk_main_iteration_do(block);
+	return 1;
 }
 
 // gtk_main_quit() may run immediately, or it may wait for other pending events; "it depends" (thanks mclasen in irc.gimp.net/#gtk+)
```

**Comment by andlabs** [comment]
Also, what would the equivalent of a `select()` be on Windows? It appears `MsgWaitForMultipleObjectsEx()` does not accept file or socket handles. Does it have to do with `OVERLAPPED`?

**Comment by andlabs** [comment]
I added uiMainSteps() now; hopefully this will allow things to work.

**Comment by andlabs** [comment]
(Copy-pasting from https://github.com/andlabs/libui/pull/21#issuecomment-485597060)

Okay, I need to resurrect this.

>It's useful for any apps where you still want control of the main loop - for instance if you have a game or gamedev tool that uses libui, and you need a tight render-loop and other updates alongside your ui. I guess you could delegate the ui handling to a separate thread, but that seems a bit of a convoluted solution (and needless threads usually cause problems).

How do people do this right now on macOS in a macOS-native application? Because there is no real supported way to do this directly in the same way libui does it now, and even if I did just recreate what `-[NSApplication run]` does internally, that wouldn't be sufficient to replace *all* of `-[NSApplication run]`.

And for the record, I am specifically referring to GUI apps here, not just programs that roll their own poll/kqueue loop.
